build_runner/lib/src/server/server.dart

```diff
@@ -330,37 +363,177 @@ class AssetHandler {
   }
 }
 
-String _renderPerformance(BuildPerformance performance, bool hideSkipped) {
+String _getPerformancePath(bool hideSkipped, bool detailedSlices,
+    int slicesResolution, PerfSortOrder sortOrder, String filter) {
+  var params = <String, String>{};
+  if (hideSkipped) params['hideSkipped'] = 'true';
+  if (detailedSlices) params['detailedSlices'] = 'true';
+  if (sortOrder.index != 0) params['sortOrder'] = '${sortOrder.index}';
+  if (slicesResolution != 5) params['slicesResolution'] = '$slicesResolution';
+  if (filter != null) params['filter'] = filter;
+  return Uri(path: '/$_performancePath', queryParameters: params).toString();
+}
+
+String _renderPerformance(
+    BuildPerformance performance,
+    bool hideSkipped,
+    bool detailedSlices,
+    int slicesResolution,
+    PerfSortOrder sortOrder,
+    String filter) {
   try {
     var rows = StringBuffer();
-    for (var action in performance.actions) {
+    final resolution = Duration(milliseconds: slicesResolution);
+    var count = 0,
+        maxSlices = 1,
+        max = 0,
+        min = performance.stopTime.millisecondsSinceEpoch -
+            performance.startTime.millisecondsSinceEpoch;
+
+    void writeRow(BuilderActionPerformance action,
+        BuilderActionStagePerformance stage, TimeSlice slice) {
+      var actionKey = '${action.builderKey}:${action.primaryInput}';
+      var tooltip = '<div class=perf-tooltip>'
+          '<p><b>Builder:</b> ${action.builderKey}</p>'
+          '<p><b>Input:</b> ${action.primaryInput}</p>'
+          '<p><b>Stage:</b> ${stage.label}</p>'
+          '<p><b>Stage time:</b> '
+          '${stage.startTime.difference(performance.startTime).inMilliseconds / 1000}s - '
+          '${stage.stopTime.difference(performance.startTime).inMilliseconds / 1000}s</p>'
+          '<p><b>Stage total duration:</b> ${stage.duration.inMilliseconds / 1000} seconds</p>'
+          '<p><b>Stage async duration:</b> ${stage.innerDuration.inMilliseconds / 1000} seconds</p>';
+      if (slice != stage) {
+        tooltip += '<p><b>Slice time:</b> '
+            '${slice.startTime.difference(performance.startTime).inMilliseconds / 1000}s - '
+            '${slice.stopTime.difference(performance.startTime).inMilliseconds / 1000}s</p>'
+            '<p><b>Slice duration:</b> ${slice.duration.inMilliseconds / 1000} seconds</p>';
+      }
+      tooltip += '</div>';
+      var start = slice.startTime.millisecondsSinceEpoch -
+          performance.startTime.millisecondsSinceEpoch;
+      var end = slice.stopTime.millisecondsSinceEpoch -
+          performance.startTime.millisecondsSinceEpoch;
+
+      if (min > start) min = start;
+      if (max < end) max = end;
+
+      rows.writeln(
+          '          ["$actionKey", "${stage.label}", "$tooltip", $start, $end],');
+      ++count;
+    }
+
+    RegExp filterRE;
```

Comment:
[optional] since we care about the _value_ not the _side effects_ I like to use the ternary conditional when possible.

```dart
final filterRegex = filter != null ? RegExp(filter) : null;
```